File: tools/bench/vii4_real_bisect_error.py
"""VII-4 본문 수치 — 실수 이분 탐색을 몇 번 돌리면 오차가 얼마가 되는가.

정확해를 아는 함수(세제곱근)로 잰다. 반복 횟수 k 를 늘리면서
(1) 남은 구간 폭, (2) 정확해와의 절대오차, (3) 상대오차를 같이 본다.
어느 지점부터 더 돌려도 오차가 줄지 않는지가 이 측정의 목적이다.

그리고 큰 값에서 절대 EPS 가 무의미해지는 것을 double 의 간격(ULP)으로 확인한다.

측정 환경은 CLAUDE.md §1-3.
실행:  python3.13 tools/bench/vii4_real_bisect_error.py
"""

import math

TARGET = 2.0
EXACT = math.cbrt(TARGET)          # 정확해. 이것과 비교한다


def cbrt_bisect(iters):
    """[0, 2) 에서 x^3 >= 2 의 경계를 iters 번 접는다."""
    lo, hi = 0.0, 2.0
    for _ in range(iters):
        mid = lo + (hi - lo) / 2
        if mid * mid * mid >= TARGET:
            hi = mid
        else:
            lo = mid
    return lo, hi


print(f"정확해 cbrt(2) = {EXACT:.17g}")
print(f"{'k':>4} {'폭':>12} {'절대오차':>12} {'상대오차':>12}")
for k in (10, 20, 30, 40, 50, 60, 70, 80, 100, 200):
    lo, hi = cbrt_bisect(k)
    width = hi - lo
    err = abs(lo - EXACT)
    rel = err / EXACT
    print(f"{k:>4} {width:12.3e} {err:12.3e} {rel:12.3e}")

# 폭이 더 줄지 않는 지점 — double 이 두 수 사이에 아무 값도 못 만든다
lo, hi = 0.0, 2.0
prev = hi - lo
stop = -1
for k in range(1, 501):
    mid = lo + (hi - lo) / 2
    if mid * mid * mid >= TARGET:
        hi = mid
    else:
        lo = mid
    if hi - lo == prev:
        stop = k
        break
    prev = hi - lo
print(f"\n폭이 더 줄지 않게 된 반복 = {stop}회, 그때의 폭 = {prev:.3e}")
print(f"1.26 근처의 double 간격(ULP) = {math.ulp(1.26):.3e}")

# 큰 값에서는 절대 EPS 가 애초에 표현 불가능하다
for x in (1.0, 1e6, 1.4e9, 1e15, 1e18):
    print(f"x = {x:9.2e} 에서 ULP = {math.ulp(x):.3e}  (EPS=1e-9 보다 "
          f"{'크다 → 절대오차 1e-9 는 불가능' if math.ulp(x) > 1e-9 else '작다 → 가능'})")

File: tools/bench/vii4_real_bisect_error.cpp
// VII-4 본문 수치 — 실수 이분 탐색을 몇 번 돌리면 오차가 얼마가 되는가 (C++ 판).
// vii4_real_bisect_error.py 와 같은 값을 내야 한다. double 연산이 IEEE 754 로
// 같기 때문에 실제로 마지막 자리까지 같다 — 그것을 확인하는 것도 이 스크립트의 몫이다.
//
// 측정 환경은 CLAUDE.md §1-3.
// 빌드/실행:
//   g++ -std=c++17 -O2 tools/bench/vii4_real_bisect_error.cpp -o /tmp/vii4 && /tmp/vii4

#include <cmath>
#include <cstdio>
using namespace std;

const double TARGET = 2.0;

void cbrt_bisect(int iters, double& lo, double& hi) {
    lo = 0.0;
    hi = 2.0;
    for (int i = 0; i < iters; i++) {
        double mid = lo + (hi - lo) / 2;
        if (mid * mid * mid >= TARGET) hi = mid;
        else lo = mid;
    }
}

// 어떤 값 x 에서 double 이 표현할 수 있는 다음 수까지의 간격
double ulp(double x) { return nextafter(x, INFINITY) - x; }

int main() {
    double exact = cbrt(TARGET);
    printf("정확해 cbrt(2) = %.17g\n", exact);
    printf("%4s %12s %12s %12s\n", "k", "폭", "절대오차", "상대오차");
    int ks[] = {10, 20, 30, 40, 50, 60, 70, 80, 100, 200};
    for (int k : ks) {
        double lo, hi;
        cbrt_bisect(k, lo, hi);
        double err = fabs(lo - exact);
        printf("%4d %12.3e %12.3e %12.3e\n", k, hi - lo, err, err / exact);
    }

    double lo = 0.0, hi = 2.0, prev = 2.0;
    int stop = -1;
    for (int k = 1; k <= 500; k++) {
        double mid = lo + (hi - lo) / 2;
        if (mid * mid * mid >= TARGET) hi = mid;
        else lo = mid;
        if (hi - lo == prev) { stop = k; break; }
        prev = hi - lo;
    }
    printf("\n폭이 더 줄지 않게 된 반복 = %d회, 그때의 폭 = %.3e\n", stop, prev);
    printf("1.26 근처의 double 간격(ULP) = %.3e\n", ulp(1.26));

    double xs[] = {1.0, 1e6, 1.4e9, 1e15, 1e18};
    for (double x : xs)
        printf("x = %9.2e 에서 ULP = %.3e  (EPS=1e-9 보다 %s)\n", x, ulp(x),
               ulp(x) > 1e-9 ? "크다 → 절대오차 1e-9 는 불가능" : "작다 → 가능");
    return 0;
}
